### PR TITLE
Add provider rebalancing zones intent

### PR DESCRIPTION
**Pull request by janedotx**
## 📚 Purpose
It's pretty similar to the permitted parking policy in that the first N-1 rules should each have only one geography, and the last one should have a maximum of 0. Any vehicle state is catchable by the first N-1 rules, but only provider drop off events should be caught by the last one.

This branch is cut off of the one adding AJV validators for policy intents.

## 👌 Resolves:
- Adds provider rebalancing intent
- 
## 🗄️ Database Changes:
none

## 💔 Breaking Changes:
- none

## ✅ PR Checklist
- [ ] Add or remove checklist items to suit your needs

### CODE REVIEW

File: packages/mds-policy-service/@types/intents/index.ts [hunk 1]
```diff
@@ -59,15 +68,18 @@ export type IntentDraft<I extends INTENT_TYPE> = {
     : I extends 'parking_time_limit'
     ? ParkingTimeLimitIntentRuleUserFields
     : I extends 'permitted_parking'
-    ? PermittedParkingIntentRuleUserFields
+    ? PermittedParkingIntentRuleUserFieldsArray
+    : I extends 'provider_rebalancing_zone'
+    ? ProviderRebalancingZoneIntentRuleUserFieldsArray
     : never
```
Comment (avatarneil) [inline packages/mds-policy-service/@types/intents/index.ts:74]: I hate typescript sometimes ☺️

File: packages/mds-policy-service/service/validators/intents.ts [hunk 2]
```diff
@@ -40,7 +47,36 @@ validatePermittedParkingAjv.addKeyword({
...
+    )
+      return false
+
+    if (!firstNMinusOneRules.every((rule: PermittedParkingIntentRuleUserFields) => rule.geographies.length === 1))
+      return false
+
+    const finalRule = rule_fields[rule_fields.length - 1]
+    if (finalRule.maximum !== 0) return false
+
+    return MICRO_MOBILITY_VEHICLE_STATES_v1_1_0.every(
```
Comment (avatarneil) [inline packages/mds-policy-service/service/validators/intents.ts:77]: Do we wanna use the `MICRO_MOBILITY_VEHICLE_STATES_v1_1_0` list directly here, or should we use the aliased `MICRO_MOBILITY_VEHICLE_STATES` array so it is automatically updated once that alias goes to a newer version?

**Review by avatarneil — APPROVED** [review]
Nice work!

**Review by avatarneil — COMMENTED** [review]
Late-breaking question, do you wanna add a changeset to this guy?

**Comment by janedotx** [comment]
> Late-breaking question, do you wanna add a changeset to this guy?

There is a changeset file.

File: packages/mds-policy-service/tests/policy-client.spec.ts [hunk 3]
```diff
@@ -379,6 +380,64 @@ describe('spot check unit test policy functions with SimplePolicy', () => {
...
+          provider_ids: [TEST_PROVIDER_ID1],
+          start_date: now(),
+          end_date: null
+        }
+      }
+
+      const { policy_id, rules } = await PolicyServiceClient.writePolicyIntentToPolicy(draft)
+      expect(rules[0]?.maximum).toEqual(10)
+      expect(rules[0]?.minimum).toEqual(0)
+      expect(rules[0]?.states).toEqual({})
```
Comment (mgeorgehansen-lacuna) [inline packages/mds-policy-service/tests/policy-client.spec.ts:424]: Might be worth checking start and end times too here

File: packages/mds-policy-service/tests/policy-client.spec.ts [hunk 4]
```diff
@@ -379,6 +380,64 @@ describe('spot check unit test policy functions with SimplePolicy', () => {
...
+        available: ['provider_drop_off'],
+        elsewhere: ['provider_drop_off'],
+        non_operational: ['provider_drop_off'],
+        on_trip: ['provider_drop_off'],
+        removed: ['provider_drop_off'],
+        reserved: ['provider_drop_off'],
+        unknown: ['provider_drop_off']
+      })
+      expect(rules[1]?.maximum).toEqual(0)
+      expect(rules[1]?.minimum).toEqual(0)
```
Comment (mgeorgehansen-lacuna) [inline packages/mds-policy-service/tests/policy-client.spec.ts:435]: May also be worth checking start and end times for the last rule too

File: packages/mds-policy-service/tests/policy-client.spec.ts [hunk 5]
```diff
@@ -379,6 +380,64 @@ describe('spot check unit test policy functions with SimplePolicy', () => {
       })
     })
 
+    it('handles publishing a ProviderRebalancingZone policy intent', async () => {
+      const draft: ProviderRebalancingZoneIntentDraft = {
+        intent_type: 'provider_rebalancing_zone',
+        rule_fields: [
+          {
```
Comment (mgeorgehansen-lacuna) [inline packages/mds-policy-service/tests/policy-client.spec.ts:387]: One potential edge case here is when there are multiple allowed rules, so might be worth adding one more rule or another test case with multiple rules, but not sure that sort of validation needs to be done at the MDS core layer (might be covered by lacuna-core so is unnecessary here).

File: packages/mds-policy-service/tests/validators.spec.ts [hunk 6]
```diff
@@ -436,4 +437,145 @@ describe('Tests Policy Intents', () => {
...
+              unknown: ['provider_drop_off']
+            }
+          }
+        ]
+      }
+
+      expect(validateProviderRebalancingZonePolicyIntent(validIntent)).not.toBeFalsy()
+    })
+
+    it('throws if the states for the last rule are wrong', () => {
```
Comment (mgeorgehansen-lacuna) [inline packages/mds-policy-service/tests/validators.spec.ts:469]: I think that this test description is wrong - the states are correct here, but the maximum is not (should equal 0).

File: packages/mds-policy-service/tests/validators.spec.ts [hunk 7]
```diff
@@ -436,4 +437,145 @@ describe('Tests Policy Intents', () => {
...
+              on_trip: ['provider_drop_off'],
+              removed: ['provider_drop_off'],
+              reserved: ['provider_drop_off'],
+              unknown: ['provider_drop_off']
+            }
+          }
+        ]
+      }
+
+      expect(() => validateProviderRebalancingZonePolicyIntent(invalidIntent)).toThrow(ValidationError)
```
Comment (mgeorgehansen-lacuna) [inline packages/mds-policy-service/tests/validators.spec.ts:492]: I'd match a regex against the validation error message to make sure that the expected validation failure is being caught. In this instance the test states that the states are wrong and we're expecting the validator to throw an error on states, but actually it's the maximum field that's wrong here and results in the validator throwing. Explicitly matching against the error message would help catch these types of issues where multiple different kinds of errors might be thrown.

File: packages/mds-policy-service/tests/validators.spec.ts [hunk 8]
```diff
@@ -436,4 +437,145 @@ describe('Tests Policy Intents', () => {
...
+    })
+
+    it('throws if the last rule does not have a max of 0', () => {
```
Comment (mgeorgehansen-lacuna) [inline packages/mds-policy-service/tests/validators.spec.ts:495]: Test description here is reversed with the one above it.

**Review by mgeorgehansen-lacuna — APPROVED** [review]
Looks good to me! There were a couple of tests where the descriptions weren't matching the expectation they were asserting against so those descriptions should be updated. Also when testing for thrown errors it's probably worth testing the error message or error code if one exists to make sure that the thrown error is the one you were expecting. But looks solid to me, TIL about ajv validators.

File: packages/mds-policy-service/tests/validators.spec.ts (hunk 7 above)
Comment (janedotx) [reply packages/mds-policy-service/tests/validators.spec.ts:492]: Sadly I couldn't figure how to get AJV to throw a nicer error than "validation for custom keyword failed" (not the exact wording of the actual AJV error, but that's about what I got). It'll require further investigating into the AJV api to figure how to get detailed error messages like "maximum < 0, this is incorrect."
